aspnetcore/mvc/razor-pages/razor-pages-convention-features.md

```diff
@@ -0,0 +1,293 @@
+* The `TryParseHandlerMethod` processes a handler to determine the HTTP verb (`httpMethod`) and named handler name (`handerName`) when creating the `PageHandlerModel`.
+  * An `Async` postfix is ignored, if present.
+  * Casing is used to parse the HTTP verb from the method name.
+  * When the method name (without `Async`) is equal to the HTTP verb name, there's no named handler. The `handerName` is set to `null`, and the method name is `Get`, `Post`, `Delete`, `Put`, or `Patch`.
+  * When the method name (without `Async`) is longer than the HTTP verb name, there's a named handler. The `handerName` is set to `<method name (less 'Async', if present)>`. For example, both "GetMessage" and "GetMessageAsync" yield a handler name of "GetMessage".
+  * DELETE, PUT, and PATCH HTTP verbs are converted to POST.
+
+Register the `CustomPageApplicationModelProvider` in the `Startup` class:
+
+[!code-csharp[Main](razor-pages-convention-features/sample/Startup.cs?name=snippet10)]
+
+The code-behind file *Index.cshtml.cs* shows how the ordinary handler method naming conventions are changed for pages in the app. The ordinary "On" prefix naming used with Razor Pages is removed. The method that initializes the page state is now named `Get`. You can see this convention used throughout the app if you open any code-behind file for any of the pages.
+
+Each of the other methods start with the HTTP verb that describes its processing. The two methods that start with `Delete` would normally be treated as DELETE HTTP verbs, but the logic in `TryParseHandlerMethod` explicitly sets the verb to POST for both handlers.
+
+Note that `Async` is optional between `DeleteAllMessages` and `DeleteMessageAsync`. They're both asynchronous methods, but you can choose to use the `Async` postfix or not; we recommend that you do. `DeleteAllMessages` is used here for demonstration purposes, but we recommend that you name such a method `DeleteAllMessagesAsync`. It doesn't affect the processing of the sample's implementation, but using the `Async` postfix calls out the fact that it's an asynchronous method.
+
+[!code-csharp[Main](razor-pages-convention-features/sample/Pages/Index.cshtml.cs?name=snippet1&highlight=1,6,16,29)]
+
+Note the handler names provided in *Index.cshtml* match the `DeleteAllMessages` and `DeleteMessageAsync` handler methods:
+
+[!code-csharp[Main](razor-pages-convention-features/sample/Pages/Index.cshtml?range=29-60&highlight=7-8,25-26)]
+
+**Additional control in CreateHandlerModel**
+
+`CreateHandlerModel` calls the `IsHandler` method ([reference source](https://github.com/aspnet/Mvc/blob/rel/2.0.1/src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageApplicationModelProvider.cs#L244-L305)) to determine which methods are handlers. You can provide your own logic that deviates from the default behavior. Just replace the `IsHandler` method with your own method at this point in the `CreateHandlerModel` method:
```

Comment:
I feel a bit iffy about documenting these. It's unlikely we'll change these, but these are methods on an `.Internal` namespace and we might change these if we want to evolve the design.

Reply:
I just meant that *they* could drop in their own if they're interested in providing an app model provider . However, that's kind'a obvious. I'll drop this whole "additional control" section.